File: core/src/main/java/com/lomekwi/cave/timeline/SegmentSet.java
package com.lomekwi.cave.timeline;

import com.lomekwi.cave.app.copy.Copyable;
import com.lomekwi.cave.app.selection.Selectable;

import java.io.Serial;
import java.io.Serializable;
import java.util.AbstractCollection;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;

/**
 * 当前选中的一组片段，实现 {@link Collection}{@code <Segment>}，
 */
public class SegmentSet extends AbstractCollection<Segment> implements Serializable, Selectable, Copyable {
    @Serial
    private static final long serialVersionUID = 1L;
    private final Set<Segment> segments = new LinkedHashSet<>();
    private transient boolean selected;

    @Override
    public boolean add(Segment segment) {
        return segments.add(segment);
    }

    @Override
    public boolean remove(Object o) {
        return segments.remove(o);
    }

    @Override
    public void clear() {
        segments.clear();
    }

    @Override
    public boolean contains(Object o) {
        return segments.contains(o);
    }

    @Override
    public boolean isEmpty() {
        return segments.isEmpty();
    }

    @Override
    public Iterator<Segment> iterator() {
        return segments.iterator();
    }

    @Override
    public int size() {
        return segments.size();
    }

    public Set<Segment> getSegments() {
        return Collections.unmodifiableSet(segments);
    }

    @Override
    public boolean isSelected() {
        return selected;
    }

    @Override
    public void setSelected(boolean selected) {
        this.selected = selected;
        for (Segment seg : segments) {
            seg.setSelected(selected);
        }
    }

    @Override
    public Copyable copy() {
        SegmentGroup commonGroup = null;
        for (Segment seg : segments) {
            SegmentGroup g = seg.getGroup();
            if (g == null) {
                commonGroup = null;
                break;
            }
            if (commonGroup == null) {
                commonGroup = g;
            } else if (g != commonGroup) {
                commonGroup = null;
                break;
            }
        }
        if (commonGroup != null) {
            return commonGroup.copy();
        }
        SegmentSet set = new SegmentSet();
        Map<SegmentGroup, SegmentGroup> groupCopies = new HashMap<>();
        for (Segment seg : segments) {
            var dup = seg.duplicate();
            dup.setTrack(seg.getTrack());
            SegmentGroup g = seg.getGroup();
            if (g != null) {
                SegmentGroup copyG = groupCopies.get(g);
                if (copyG == null) {
                    copyG = new SegmentGroup();
                    groupCopies.put(g, copyG);
                }
                copyG.add(dup);
            }
            set.add(dup);
        }
        return set;
    }
}

File: core/src/main/scala/com/lomekwi/cave/timeline/SegmentSet.scala
package com.lomekwi.cave.timeline

import com.lomekwi.cave.app.copy.Copyable
import com.lomekwi.cave.app.selection.Selectable

import java.io.Serializable
import java.util
import java.util.{Collections}

import scala.jdk.CollectionConverters.*

/**
 * 当前选中的一组片段，实现 {@link Collection}{@code <Segment>}，
 */
@SerialVersionUID(1L)
class SegmentSet extends util.AbstractCollection[Segment] with Serializable with Selectable with Copyable {
  private final val segments: util.Set[Segment] = new util.LinkedHashSet[Segment]()
  @transient private var selected: Boolean = false

  override def add(segment: Segment): Boolean = {
    segments.add(segment)
  }

  override def remove(o: Any): Boolean = {
    segments.remove(o)
  }

  override def clear(): Unit = {
    segments.clear()
  }

  override def contains(o: Any): Boolean = {
    segments.contains(o)
  }

  override def isEmpty: Boolean = {
    segments.isEmpty
  }

  override def iterator(): util.Iterator[Segment] = {
    segments.iterator()
  }

  override def size(): Int = {
    segments.size()
  }

  def getSegments: util.Set[Segment] = {
    Collections.unmodifiableSet(segments)
  }

  override def isSelected: Boolean = {
    selected
  }

  override def setSelected(selected: Boolean): Unit = {
    this.selected = selected
    for (seg <- segments.asScala) {
      seg.setSelected(selected)
    }
  }

  override def copy(): Copyable = {
    val groups = segments.asScala.toSeq.map(seg => Option(seg.getGroup))
    val commonGroup = groups.headOption.flatten.filter(g => groups.forall(_.contains(g)))

    commonGroup.map(g => g.copy()).getOrElse {
      val set = new SegmentSet()
      val groupCopies: util.Map[SegmentGroup, SegmentGroup] = new util.HashMap[SegmentGroup, SegmentGroup]()
      for (seg <- segments.asScala) {
        val dup = seg.duplicate()
        dup.setTrack(seg.getTrack)
        val g = seg.getGroup
        if (g != null) {
          var copyG = groupCopies.get(g)
          if (copyG == null) {
            copyG = new SegmentGroup()
            groupCopies.put(g, copyG)
          }
          copyG.add(dup)
        }
        set.add(dup)
      }
      set
    }
  }
}
